src/main/java/com/redhat/devtools/lsp4ij/client/features/LSPFormattingFeature.java

```diff
@@ -0,0 +1,93 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at https://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.lsp4ij.client.features;
+
+import com.intellij.lang.LanguageFormatting;
+import com.intellij.psi.PsiFile;
+import com.redhat.devtools.lsp4ij.LanguageServerItem;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * LSP formatting feature.
+ */
+@ApiStatus.Experimental
+public class LSPFormattingFeature extends AbstractLSPFeature<LSPFormattingFeature.LSPFormattingFeatureContext> {
+
+    public static class LSPFormattingFeatureContext {
+
+        private final @NotNull PsiFile file;
+        private final @NotNull LSPFormattingFeature.LSPFormattingFeatureContext.@NotNull FormattingKind formattingKind;
+
+        public enum FormattingKind {
+            ONLY_DOCUMENT,
+            DOCUMENT_AND_RANGE;
+        }
+
+        public LSPFormattingFeatureContext(@NotNull PsiFile file, @NotNull FormattingKind formattingKind) {
+            this.file = file;
+            this.formattingKind = formattingKind;
+        }
+
+        public @NotNull PsiFile getFile() {
+            return file;
+        }
+
+        public @NotNull LSPFormattingFeature.LSPFormattingFeatureContext.@NotNull FormattingKind getFormattingKind() {
+            return formattingKind;
+        }
+    }
+
+    @Override
+    public boolean isEnabled(@NotNull LSPFormattingFeatureContext featureContext) {
+        PsiFile file = featureContext.getFile();
+        if (!canFormatEvenIfExistingFormatter(file)) {
+            return false;
+        }
+        if (featureContext.getFormattingKind() == LSPFormattingFeatureContext.FormattingKind.DOCUMENT_AND_RANGE) {
+            return isDocumentRangeFormattingSupported(file);
+        }
+        return isDocumentFormattingSupported(file) && !isDocumentRangeFormattingSupported(file);
+    }
+
+    /**
+     * Returns true if LSP formatting can be done even if it exists a custom formatter and false otherwise.
+     *
+     * @param file the file.
+     * @return true if LSP formatting can be done even if it exists a custom formatter and false otherwise.
+     */
+    protected boolean canFormatEvenIfExistingFormatter(@NotNull PsiFile file) {

```

Comment:
We have dicussed with @fbricon and we decide to have an `isEnabled `method for each `LSP*Feature` with a custom context instead of having isEnabled with several parameters to be sure that API will not broken if we need new fields.

> According to the doc comments, ideCanFormatThisFileItself means whether LanguageFormatting.forContext return a non-null value

Thanks for your feedback. I removed this canFormatEvenIfExistingFormatter and now you have now:

```java
    /**
     * Returns true if existing formatter are overrideable and false (default value) otherwise.
     *
     * @return true if existing formatter are overrideable and false (default value) otherwise.
     */
    protected boolean isExistingFormatterOverrideable(@NotNull PsiFile file) {
        return false;
    }
```

In this case `LanguageFormatting.INSTANCE.forContext(file) != null` i sdone.

If you override isExistingFormatterOverrideable to returns true, it will skip the test with LanguageFormatting.INSTANCE.forContext(file) != null`

@InSyncWithFoo @FalsePattern @CppCXY do you like this idea?

Reply:
> means whether the file matches the registered DocumentSelectors.

Today we checkserver capability, but not DocumentSelectors. Once LSP API will be merged,I will work on https://github.com/redhat-developer/lsp4ij/issues/517